src/lib.rs

```diff
@@ -0,0 +1,420 @@
+#![cfg_attr(not(feature = "std"), no_std)]
+
+use codec::{Codec, Decode, Encode};
+use rstd::collections::btree_set::BTreeSet;
+use rstd::prelude::*;
+use runtime_primitives::traits::{MaybeSerialize, Member, One, SimpleArithmetic};
+use srml_support::{
+    decl_event, decl_module, decl_storage, dispatch, ensure, traits::Get, Parameter, StorageMap,
+    StorageValue,
+};
+use system::{self, ensure_signed};
+
+mod error_messages;
+mod mock;
+mod tests;
+
+use error_messages::*;
+
+type MaxLength = u32;
+
+type MaxNumber = u32;
+
+type MaxConsecutiveRepliesNumber = u16;
+
+/// The pallet's configuration trait.
+pub trait Trait: system::Trait {
+
+    /// The overarching event type.
+    type Event: From<Event<Self>> + Into<<Self as system::Trait>::Event>;
+
+    //Security/configuration constraints
+    type PostTitleMaxLength: Get<MaxLength>;
+
+    type PostBodyMaxLength: Get<MaxLength>;
+
+    type ReplyMaxLength: Get<MaxLength>;
+
+    type PostsMaxNumber: Get<MaxNumber>;
+
+    type RepliesMaxNumber: Get<MaxNumber>;
+
+    type DirectRepliesMaxNumber: Get<MaxNumber>;
+
+    type ConsecutiveRepliesMaxNumber: Get<MaxConsecutiveRepliesNumber>;
+
+    /// Type of identifier for blogs.
+    type BlogId: Parameter
+        + Member
+        + SimpleArithmetic
+        + Codec
+        + Default
+        + Copy
+        + MaybeSerialize
+        + PartialEq;
+
+    /// Type of identifier for blog posts.
+    type PostId: Parameter
+        + Member
+        + SimpleArithmetic
+        + Codec
+        + Default
+        + Copy
+        + MaybeSerialize
+        + PartialEq;
+
+    /// Type of identifier for replies.
+    type ReplyId: Parameter
+        + Member
+        + SimpleArithmetic
+        + Codec
+        + Default
+        + Copy
+        + MaybeSerialize
+        + PartialEq;
+}
+
+#[derive(Encode, Decode, Default, Clone, PartialEq)]
+#[cfg_attr(feature = "std", derive(Debug))]
+pub struct Post {
+    title: Vec<u8>,
+    body: Vec<u8>,
+}
+
+impl Post {
+    fn new(title: Vec<u8>, body: Vec<u8>) -> Self {
+        Self { title, body }
+    }
+}
+
+// Blog`s pallet storage items.
+decl_storage! {
+    trait Store for Module<T: Trait> as BlogModule {
+
+        // Blog Ids set, associated with owner
+        BlogIds get(fn blog_ids_by_owner): map T::AccountId => Option<BTreeSet<T::BlogId>>;
+
+        // Post Ids set, associated with blog
+        BlogPostIds get(fn post_ids_by_blog_id): map T::BlogId => Option<BTreeSet<T::PostId>>;
+
+        // Root replies
+        BlogPostParentReplyIds get (fn post_root_reply_ids): map (T::BlogId, T::PostId) => Option<BTreeSet<T::ReplyId>>;
+
+        // Direct replies to reply
+        BlogPostChildReplyIds get (fn post_child_reply_ids): map (T::BlogId, T::PostId, T::ReplyId) => Option<BTreeSet<T::ReplyId>>;
+
+        BlogPost get(fn post_by_id): map (T::BlogId, T::PostId) => Option<Post>;
+
+        Reply get (fn reply_by_id): map (T::BlogId, T::PostId, T::ReplyId) => Option<Vec<u8>>;
```

Comment:
- Rename to `ReplyById`.

- Introduce a proper `Reply` struct, not this raw content representation.

- Why is the value type an `Option`?

- Make value in first dimension efficiently deletable by doing `double_map T::BlogId, (T::PostId, T::ReplyId) => _ `

Reply:
- Ok.
-  Ok.
-  In case of key pair non existence we will get none instead of default case for this value. Though we can check existence before accessing, using corresponding method, getter will return default representation.
- Ok

Reply:
> * Make value in first dimension efficiently deletable by doing `double_map T::BlogId, (T::PostId, T::ReplyId) => _`

Behedo, seems we have an issue, using double_map in `rev c37bb08535c49a12320af7facfd555ce05cce2e8` To be honest, I didn`t find double_map in any of our runtime modules.
I also tried to run double_map pallet example with substrate version, used in our runtime modules, getting same error. https://github.com/substrate-developer-hub/recipes/tree/master/pallets/double-map
 error: expected `;`
  --> pallets/double-map/src/lib.rs:32:15
   |
32 |             double_map hasher(blake2_256) GroupIndex, hasher(blake2_256) T::AccountId => u32;
   |                        ^^^^^^
Seems, that we need update our substrate version to be able to use double_map

Reply:
It should work fine, here is it is being used in our forum module

https://github.com/Joystream/substrate-forum-module/blob/development/forum-module/src/lib.rs#L736

Reply:
> It should work fine, here is it is being used in our forum module
> 
> https://github.com/Joystream/substrate-forum-module/blob/development/forum-module/src/lib.rs#L736

I am not sure sure, what the reason is, but here we are using another substrate version, 0e3001a1ad6fa3d1ba7da7342a8d0d3b3facb2f3
In our monorepo, forum module is currently using `rev c37bb08535c49a12320af7facfd555ce05cce2e8` substrate and doesn`t have `double_map` as well.

Reply:
You can see that double_map is indeed part of this version of Substrate

https://github.com/paritytech/substrate/blob/c37bb08535c49a12320af7facfd555ce05cce2e8/srml/session/src/lib.rs#L404

I suggest you go to the Substrate Matrix dev chat to get assistance if you get totally blocked here.

Reply:
You are right, but syntax changed a little bit. `get(some fn):` macro syntax before `double_map`  is forbidden in our substrate version.